dashboard/app/models/pd/workshop.rb

```diff
@@ -495,10 +495,37 @@ def process_location
     self.processed_location = {
       latitude: result.latitude,
       longitude: result.longitude,
+      city: result.city,
+      state: result.state,
       formatted_address: result.formatted_address
     }.to_json
   end
 
+  # helper methods for retrieving processed location data, with a possible side
+  # effect of reprocessing the data. Useful for fields newly-added to process
+  # location which may or may not be defined on older data
+  def location_city
+    location_hash = JSON.parse processed_location
+    unless location_hash.key? 'city'
+      process_location
+      update_column :processed_location, processed_location
+      location_hash = JSON.parse processed_location
+    end
+
+    location_hash['city']
+  end
+
+  def location_state
+    location_hash = JSON.parse processed_location
+    unless location_hash.key? 'state'
+      process_location
+      update_column :processed_location, processed_location
+      location_hash = JSON.parse processed_location
+    end
```

Comment:
Now that I see this written out, it looks like we can combine the common part into something like

```ruby
def get_processed_location_part(key)
  location_hash = JSON.parse processed_location
  unless location_hash.key? key.to_s
    process_location
    update_column :processed_location, processed_location
    location_hash = JSON.parse processed_location
  end

  location_hash[key]
end
```

Then:
```ruby
def location_city
  get_processed_location_part :city
end
```

Reply:
I'm conflicted about this.

In part because I feel like the introduction of a reprocess should in some sense be a temporary addition just to deal with the fact that this is a new field, and should be removed anyway once we can be confident all our data is up-to-date.

But also in part because with the specific methods we have some degree of confidence that if the hash doesn't contain `city` before, it will afterwards, but if we call `get_processed_location_part("some invalid field")` we could end up making a ton of useless api requests.

Thoughts?

Reply:
Good points. If it's to be temporary, let's add a comment.

We could instead run a oneoff script to reprocess location on all upcoming workshops, and leave this out.

If we leave it in, I'm also conflicted - I'd rather not have that duplicate code, bu I'd even more rather not leave the option of endless useless api requests. We could restrict the keys to only `%w(city state)` and otherwise just return nil.